README.md
# EEC-MGC

Practicals for Master module "Macroecology and global change" at Univ. Potsdam: https://damariszurell.github.io/EEC-MGC/


old_scripts/changes.txt
Major changes:

folder 2022: major switch from raster to terra
- scripts before changing to terra for most practicals
- more elaborate pseudo-absence practical, with IDW, largely building on disco
- bulk rasterizing range maps still largely based on letsR

folder 2023: switch from dismo to predicts
- remove DOMAIN
- dismo is now only used for gbm.step 


Macro_terra_version/MGC_a1_SpatialData.R
library(sf)
library(terra)

# old:
# library(raster)
# library(sp)
# library(dismo)

rm(list = ls())

###############

# 2.1 Spatial
# We set a seed for the random number generator, so that we all obtain the same results
set.seed(12345)

coords <- cbind(
  x <- rnorm(10, sd=2),
  y <- rnorm(10, sd=2)
)

str(coords)
#-------------------------------------------------------------------------------

plot(coords)

#-------------------------------------------------------------------------------

# # Convert into SpatialPoints
# old:
# sp <- sp::SpatialPoints(coords)

# ISSUE: you can't use coords directly with sf. 
# terra::vect works fine if you just want a print summary similar to raster (no information on the bounding box)
# for a similar output to showDefault(sp): convert sp to sn sf object using st_as_sf

# new version
sp <- terra::vect(coords)

# Check out the object
class(sp)


# OPTIONAL ########
# sp <- st_as_sf(sp)
# class(sp)
# showDefault(sp) # This holds way less information than an SpatialPoints object from the sp package
##################

#-------------------------------------------------------------------------------

# old
# sp <- SpatialPoints(coords, proj4string=CRS('+proj=longlat +datum=WGS84'))
# sp

# terra version
sp <- terra::vect(coords, crs = '+proj=longlat +datum=WGS84')
sp

#-------------------------------------------------------------------------------

# Create attribute table
(data <- data.frame(ID=1:10,species=sample(c('beech','oak','birch'),10,replace=T)))

#-------------------------------------------------------------------------------

# old
# # Create SpatialPointsDataFrame
# (spdf <- SpatialPointsDataFrame(sp, data))

# for terra, you can simply add the additional information to the existing Spatial Vector
sp_trees <- sp
(terra:: values(sp_trees) <- data)


# alternative: combine everything right away
# (sp_trees2 <- terra::vect(coords, crs = '+proj=longlat +datum=WGS84', atts = data))
  

########################################
# 2.2 SpatialLines and SpatialPolygons #
########################################

# Create SpatialLines through all oak trees in the data

#old:
# lns  <- raster::spLines(subset(spdf,species=='oak'), crs=CRS('+proj=longlat +datum=WGS84'))
# pols <- raster::spPolygons(subset(spdf,species=='birch'), crs=CRS('+proj=longlat +datum=WGS84'))

# prepare subsets for the two tree species 
oak   <- terra::subset(sp_trees, sp_trees$species == 'oak')
birch <- terra::subset( sp_trees, sp_trees$species == 'birch')

# sp_oak with the geometry 'points' can be plotted as lines as it is, but sp_birch needs to be modified before we can plot it as polygon:
# get the coordinates for the birch trees and transform them to a SpatialVector (type = polygons)
birch_coords <- terra::crds(birch)
birch <- terra::vect(birch_coords, type = "polygons")

plot(sp, axes = TRUE, lwd = 2)
terra::lines(oak, col = 'red', lwd = 2)
terra::polys(birch, border = 'blue', col = 'yellow', lwd = 2, alpha = 1)

#####################################
# 2.3 Reading vector data from file #
#####################################
# old:
#(shrew <- raster::shapefile('data/IUCN_Sorex_alpinus.shp'))

# terra:
(shrew <- terra::vect('data/IUCN_Sorex_alpinus.shp'))

#-------------------------------------------------------------------------------

# Plot Central Europe
library(maps)
map('world', xlim = c(5,30), ylim = c(40,55))

# overlay the range of the Alpine Shrew
plot(shrew, col = "red", add = TRUE)

# or:
# terra::polys(shrew, col = "red")
#-------------------------------------------------------------------------------

######################
# 3 Raster data in R #
######################

# old:
# (r1 <- raster(ncol=10, nrow=10, xmx=-80, xmn=-150, ymn=20, ymx=60))

(r1 <- terra::rast(ncol = 10, nrow = 10, xmax = -80, xmin = -150, ymax = 60, ymin = 20 ))

#-------------------------------------------------------------------------------
# old:
# summary(raster::values(r1))
# raster::values(r1) <- rnorm(ncell(r1))


summary(terra::values(r1))

terra::values(r1) <- rnorm(ncell(r1))

# plot the raster
plot(r1)

#-------------------------------------------------------------------------------

# create another raster layer and assign values
r2 <- r1

# old:
# raster::values(r1) <- rnorm(ncell(r1))
terra::values(r2) <- 1:ncell(r2)

# Stack the raster layers
# old:
# (r <- stack(r1,r2))
(r <- c(r1, r2)) # c() as alternative to stack()

#-------------------------------------------------------------------------------

plot(r)

#-------------------------------------------------------------------------------

###########################
# 3.1 Read in raster data #
###########################

# old:
# (temp <- raster('data/UK_temp.tif'))
(temp <- terra::rast('data/UK_temp.tif'))

# ------------------------------------------------------------------------------

plot(temp)

#-------------------------------------------------------------------------------

# old:
# (bioclim <- stack('data/UK_bioclim.grd'))
(bioclim <- terra::rast('data/UK_bioclim.grd')) # as alternative to stack()

#-------------------------------------------------------------------------------

plot(bioclim)

#-------------------------------------------------------------------------------

############################
# 3.2 Download raster data #
############################

# old:
# (clim <- getData("worldclim", var="bio", res=10, download=T, path="data"))

library(geodata) 

(clim <- geodata::worldclim_global(var = 'bio', res = 10, download = T, path = 'data')) # alternative to getData()

plot(clim)

##############################
# 3.3 Manipulate raster data #
##############################

# old: (examples)
# temp_eur <- crop(clim[[1]], c(-15,45,35,72))
# temp_eur_onedeg <- aggregate(temp_eur, 6)


# Crop the temperature layer (bio1) to roughly European extent
temp_eur <- terra::crop(clim[[1]], c(-15,45,35,72))

# Aggregate to one-degree and two-degree resolution
temp_eur_onedeg <- terra::aggregate(temp_eur, 6)
temp_eur_twodeg <- terra::aggregate(temp_eur, 12)

par(mfrow=c(1,3))
plot(temp_eur)
plot(temp_eur_onedeg)
plot(temp_eur_twodeg)

###########################
# 3.4 Extract raster data #
###########################

# Generate random locations
# old:
# library(dismo)
# lonlat <- randomPoints(temp_eur, 10)

lonlat <- terra::spatSample(temp_eur, size = 10, method = 'random', replace = FALSE, xy = TRUE, values = FALSE) # xy to return coordinates

# dev.off() to reset par()

# Map temperature and locations
plot(temp_eur)
points(lonlat, cex=2, pch=19)

#------------------------------------------------------------------------------- 

# Extract temperature values at these locations
# old:
#raster::extract(temp_eur, lonlat)

terra::extract(temp_eur, lonlat)


Macro_terra_version/MGC_a6_BiodivChanges.R
library(terra)

rm(list = ls())

# Read in the distribution dataset:
bird_dist <- read.table('data/UK_BBatlas_1968_2008.csv',header=T, sep=',', stringsAsFactors = F)

# Look at data, the first three columns contain coordinate information, the 4th column contains the time periods:
head(bird_dist[,1:20])

#-------------------------------------------------------------------------------

# Species names are contained in the remaining columns:
spp <- names(bird_dist)[-c(1:4)]
spp_names <- sub('_',' ',spp)

#-------------------------------------------------------------------------------

###############################
# Species richness changes ####
###############################

# Extract species occurrences per time period and compute species richness:
birds_68 <- subset(bird_dist,period=='1968-72')
birds_68$richness68 <- rowSums(birds_68[,spp])

# old:
# plot(rasterFromXYZ(birds_68[,c('EASTING','NORTHING','richness68')]))

plot(terra::rast(birds_68[, c('EASTING', 'NORTHING', 'richness68')], type = 'xyz'))

#-------------------------------------------------------------------------------

# Calculate species richness
bird_dist$richness <- rowSums(bird_dist[,spp])


library(raster)

# !!!
# NOTE: this has to be done with spplot/raster to achieve the desired plot. 
# !!!
# Map species richness for different time periods
spplot(stack(
  rasterFromXYZ(subset(bird_dist,period=='1968-72')[,c('EASTING','NORTHING','richness')]),
  rasterFromXYZ(subset(bird_dist,period=='1988-91')[,c('EASTING','NORTHING','richness')]),
  rasterFromXYZ(subset(bird_dist,period=='2008-11')[,c('EASTING','NORTHING','richness')])),
  names.attr = c('1968-72','1988-91','2008-11'), main='Species richness'
)

#-------------------------------------------------------------------------------

#########################################
# 2 Mapping species richness changes ####
#########################################

# Define axis for colour breaks
brks_diff <- seq(-50,50,by = 5)

# Define colour palette
cols.diff = colorRampPalette(
  rev(c('#650A13','#b2182b','#d6604d','#f4a582','grey90','#92c5de','#4393c3','#2166ac','#0B2F52')))(21) # red-blue diverging

# Calculate the difference in species richness between 1968 and 2011
# old
# change_richness <- rasterFromXYZ(subset(bird_dist,period=='2008-11')[,c('EASTING','NORTHING','richness')]) -
#   rasterFromXYZ(subset(bird_dist,period=='1968-72')[,c('EASTING','NORTHING','richness')])

change_richness <- terra::rast(subset(bird_dist,period=='2008-11')[,c('EASTING','NORTHING','richness')]) -
  terra::rast(subset(bird_dist,period=='1968-72')[,c('EASTING','NORTHING','richness')])

# !!!
# NOTE: original code in the practical defines the object brks_diff but then uses brks.diff in the plot function
# !!!
# Plot the species richness change between 1968 and 2011
plot(change_richness, axes=F,  col=cols.diff, lab.breaks=brks_diff, zlim=c(-50,50), 
     main='Species richness change \n1968 - 2011')

#-------------------------------------------------------------------------------

##########################################################
# 2.1 Mapping taxon-specific species richness changes ####
##########################################################

# !!!
# TODO: remove one of the two code boxes where redlist_status is defined.
# !!!
redlist_status <- read.table('data/UK_birds_redlist_status.csv', header=T, sep=',')

#-------------------------------------------------------------------------------

# Restrict red list table to those species that are found in UK:
redlist_status <- subset(redlist_status, scientific_name %in% spp_names)

# List number of bird species in different orders:
table(redlist_status$order)

#-------------------------------------------------------------------------------

# List number of bird species in different families:
table(redlist_status$family)

#-------------------------------------------------------------------------------

# List bird species in order ACCIPITRIFORMES:
subset(redlist_status,order=='ACCIPITRIFORMES')$scientific_name

#-------------------------------------------------------------------------------

# Extract species belonging to the order ACCIPITRIFORMES:
accipiter_spp <- subset(redlist_status,order=='ACCIPITRIFORMES')$scientific_name
accipiter_spp <- sub(' ','_', accipiter_spp)

# Compute ACCIPITRIFORMES species richness
birds_68$sr_accipiter <- rowSums(birds_68[,accipiter_spp])

# Map ACCIPITRIFORMES species richness
# old
# plot(rasterFromXYZ(birds_68[,c('EASTING','NORTHING','sr_accipiter')]), main="Richness ACCIPITRIFORMES")

plot(terra::rast(birds_68[,c('EASTING','NORTHING','sr_accipiter')], type = 'xyz'), main = 'Richness ACCIPITRIFORMES')

#-------------------------------------------------------------------------------

#############################################
# 3.1. Quantifying species range changes ####
#############################################

# no changes required in this section

#-------------------------------------------------------------------------------

########################################
# 3.2 Mapping species range changes ####
########################################

# !!!
# NOTE: raster is required here to be able to use spplot
# !!!

# ranges sedge warbler
sedge_68 <- rasterFromXYZ(subset(bird_dist,period=='1968-72')[,c('EASTING','NORTHING','Acrocephalus_schoenobaenus')])
sedge_88 <- rasterFromXYZ(subset(bird_dist,period=='1988-91')[,c('EASTING','NORTHING','Acrocephalus_schoenobaenus')])
sedge_08 <- rasterFromXYZ(subset(bird_dist,period=='2008-11')[,c('EASTING','NORTHING','Acrocephalus_schoenobaenus')])

# Plot the range in different time periods
spplot(stack(sedge_68, sedge_88, sedge_08))

#-------------------------------------------------------------------------------

# !!!
# keep on working with the sedge objects that were created using raster
# !!!

# Compute differences. To distinguish clearly where cells are colonised or decolonised, we assign occurrences in 1968 a value=1 and occurrence in 2008 a value=2:
dummy <- sedge_08
values(dummy)[values(dummy)>0 & !is.na(values(dummy))] <- 2

# Sum up the layers. Values=1 indicate presence in 1968, values=2 presence in 2008, values=3 presence in both periods
diff <- sedge_68+dummy
diff_f <- as.factor(diff)
rat <- levels(diff_f)[[1]]
rat[["diff"]] <- c("Absent", "Extinction", "Colonisation", "Stable")
levels(diff_f) <- rat

library(rasterVis)
custom.pal <- c("grey75", "coral", "cyan3","grey25")
levelplot(diff_f, margin=F, scales=list(draw=FALSE),col.regions=custom.pal,main='Sedge warbler range change (1968-2011)')



Macro_terra_version/MGC_a5_SpeciesThreats.R
library(terra)

rm(list = ls())

# Read in the distribution dataset:
bird_dist <- read.table('data/UK_BBatlas_2008.csv', header = TRUE, sep = ',', stringsAsFactors = FALSE)

# Species names are contained in the columns:
spp <- names(bird_dist)[-c(1:3)]

# For later usage, we need to remove the underscore in the names:
spp <- sub('_',' ',spp)

redlist_status <- read.table('data/UK_birds_redlist_status.csv', header=T, sep=',')

#-------------------------------------------------------------------------------

# !!!
# I skipped the first section where raster was not used


##############################################
# 1.2 Mapping hotspots of threatened species #
##############################################

library(raster)

# !!!
# NOTE: raster required here to be able to use spplot
# !!!

# Identify all vulnerable species 
vu_spp <- subset(redlist_status,category=='VU')$scientific_name
# We have to make sure that species names are written in the same way in the redlist and distribution data
vu_spp <- sub(' ','_', intersect(vu_spp,spp))

# Identify all least concern species
lc_spp <- subset(redlist_status,category=='LC')$scientific_name
lc_spp <- sub(' ','_', intersect(lc_spp,spp))

# Now, we extract the distribution data for the VU and LC species groups, make rasters, stack these and plot
spplot(stack(
  rasterFromXYZ(data.frame(bird_dist[,2:3],log_VU=log(rowSums(bird_dist[,vu_spp])))),
  rasterFromXYZ(data.frame(bird_dist[,2:3],log_LC=log(rowSums(bird_dist[,lc_spp]))))), 
  main='log( species richness)')

#-------------------------------------------------------------------------------

#########################
# 2 Red list threats ####
#########################

redlist_threats <- read.table('data/UK_birds_redlist_threats.csv', header = TRUE, sep = ',')

#-------------------------------------------------------------------------------

redlist_threats[sample(nrow(redlist_threats),10),-c(1:2)]

#-------------------------------------------------------------------------------

table(redlist_threats$timing)

#-------------------------------------------------------------------------------

######################################
# 2.1 Mapping hotspots of threats ####
######################################

# Which ongoing threats are the most common ?
sort(table(subset(redlist_threats, species %in% spp & timing=='Ongoing')$title), decreasing=T)[1:10]

#-------------------------------------------------------------------------------

# Identify the species experiencing threats from hunting
spp_threat1 <- sub(' ','_',subset(redlist_threats,title=="Hunting & trapping terrestrial animals" & species %in% spp)$species)
# Identify the species experiencing threats from industry and military
spp_threat2 <- sub(' ','_',subset(redlist_threats,title=="Industrial & military effluents" & species %in% spp)$species)

# Map species experiencing threats from hunting
plot(terra::rast(data.frame(bird_dist[,2:3],rowSums(bird_dist[,spp_threat1])), type = 'xyz'), main="Hunting & trapping terrestrial animals")

#-------------------------------------------------------------------------------

# species experiencing threats from industry and military
plot(terra::rast(data.frame(bird_dist[,2:3],rowSums(bird_dist[,spp_threat2])), type = 'xyz'), main="Industrial & military effluents")



Macro_terra_version/MGC_a2_RichnessGradients.R
# library(raster)
library(terra)

rm(list = ls())

# Read in the distribution dataset:
bird_dist <- read.table('data/UK_BBatlas_2008.csv',header=T, sep=',', stringsAsFactors = F)

str(bird_dist, list.len = 10)

#-------------------------------------------------------------------------------

############################
# 1.1 Distribution data ####
############################

library(terra)

# The coordinate reference system for the UK National Grid is:
proj_UK <- '+proj=tmerc +lat_0=49 +lon_0=-2 +k=0.9996012717 +x_0=400000 +y_0=-100000 +ellps=airy +datum=OSGB36 +units=m +no_defs'

# old:
# r_ptarmigan <- rasterFromXYZ(bird_dist[,c('EASTING','NORTHING','Lagopus_lagopus')], crs=proj_UK)

# terra version 
r_ptarmigan <- terra::rast(bird_dist[,c('EASTING','NORTHING','Lagopus_lagopus')], crs=proj_UK, type = "xyz")


# terra plot will look different than the raster version 
plot(r_ptarmigan)

#-------------------------------------------------------------------------------

####################################
# 1.2 Species richness patterns ####
####################################

richness <- data.frame(bird_dist[,2:3], richness = rowSums(bird_dist[,-c(1:3)]))
head(richness)

# old:
# r_richness <- rasterFromXYZ(richness, crs=proj_UK)

# terra version
r_richness <- terra::rast(richness, crs = proj_UK, type = "xyz")

plot(r_richness)

#-------------------------------------------------------------------------------

########################################################
# 1.3 Task 2: Latitudinal species richness gradient ####
########################################################

library(ggplot2)
ggplot(data = richness, mapping = aes(x = NORTHING, y = richness)) + 
  geom_point() + 
  geom_smooth(method="loess") + 
  xlab('Latitude (UK NORTHING)') + 
  ylab('Species richness')

#-------------------------------------------------------------------------------

#######################################
# 1.4 Species-energy relationships ####
#######################################

# old:
# clim <- stack('data/UK_bioclim.grd')

clim <- terra::rast('data/UK_bioclim.grd')

plot(clim)

# TODO: !!! it appears that one of the code boxes in the practical is missing (the one where richness2 is created)

# !!! my version to create richness2
richness2 <- cbind(richness, terra::extract(clim, richness[,1:2]))


# save(richness2,file='data/UK_bird_richness_clim.RData')

#-------------------------------------------------------------------------------

##################################
# 2 -  More richness patterns ####
##################################

# Read in the distribution dataset:
bird_dist2 <- read.table('data/UK_BBatlas_1968_2008.csv',header=T, sep=',', stringsAsFactors = F)

# Look at data, the first three columns contain coordinate information, the 4th column contains the time periods:
head(bird_dist2[,1:20])

#-------------------------------------------------------------------------------

# Species names are contained in the remaining columns:
spp <- base::names(bird_dist2)[-c(1:4)]
spp_names <- sub('_',' ',spp)

#-------------------------------------------------------------------------------

# Which time periods are contained in the data?
unique(bird_dist2$period)

#-------------------------------------------------------------------------------

# Extract species occurrences per time period and compute species richness:
birds_68 <- base::subset(bird_dist2, period == '1968-72')
birds_68$richness68 <- rowSums(birds_68[,spp])

# old:
# plot(rasterFromXYZ(birds_68[,c('EASTING','NORTHING','richness68')]))
plot(terra::rast(birds_68[,c('EASTING','NORTHING','richness68')], type = "xyz"))


Macro_terra_version/MGC_a4_RangeMaps.R
# library(raster)
# library(sf)

library(terra)


rm(list = ls())

##########################
# 1.1 IUCN range maps ####
##########################

# Load the shapefile
# old: 
# (shrew <- raster::shapefile('data/IUCN_Sorex_alpinus.shp'))

(shrew <- terra::vect('data/IUCN_Sorex_alpinus.shp'))

#-------------------------------------------------------------------------------

# Plot the Central Europe
library(maps)
map('world',xlim = c(5,30), ylim = c(40,55))

# Overlay the range of the Alpine Shrew
plot(shrew, col = 'red', add = TRUE)

#-------------------------------------------------------------------------------
# Read shapefile for all mammals using the raster package:
# old:
# mammals <- raster::shapefile('data/MAMMTERR.shp')
# raster::head(mammals)

mammals <- terra::vect('data/MAMMTERR.shp')

#-------------------------------------------------------------------------------

mammals

#-------------------------------------------------------------------------------

# Inspect attribute table
terra::head(mammals)

#-------------------------------------------------------------------------------

# Show all entries for the species 'Lynx lynx'
# old:
# raster::subset(mammals_old@data, BINOMIAL=='Lynx lynx')

terra::values(terra::subset(mammals, mammals$BINOMIAL == "Lynx lynx"))
# !!!
# just using terra::subset doesn't return the entries
# !!!


# Show all entries for the species with the word 'Lynx' in their name
# old:
# grep('Lynx', mammals_old@data$BINOMIAL)
# mammals_old@data[grep('Lynx',mammals_old@data$BINOMIAL),]


grep('Lynx', mammals$BINOMIAL)
mammals[grep('Lynx', mammals$BINOMIAL),] # !!! output differs from raster output

#-------------------------------------------------------------------------------

# Range map of the Eurasian lynx
# old:
# lynx_lynx_old <- mammals_old[mammals_old@data$BINOMIAL=='Lynx lynx',]

lynx_lynx <- mammals[mammals$BINOMIAL=='Lynx lynx',]


# Map the range
library(maps)
map('world')
terra::plot(lynx_lynx, col='red', add=T)

#-------------------------------------------------------------------------------
##########################
# 1.2 BIEN range maps ####
##########################

library(BIEN) # depends on raster

# old:
# (monkey_puzzle <- BIEN_ranges_load_species('Araucaria_araucana'))

# IMPORTANT: BIEN loads the data in as SpatialPolygonsDataFrame. We just change it right away.
# This might cause problems in once sp is retired

(monkey_puzzle <- terra::vect(BIEN_ranges_load_species('Araucaria_araucana')))
# there will be a warning message about the crs (always appears when directly
# transforming a SpatialPolygonsDataFrame with vect())

# this solves the "issue":
crs(monkey_puzzle) <- '+proj=longlat +datum=WGS84 +no_defs'

#-------------------------------------------------------------------------------

# Map
map('world', xlim = c(-180,-20), ylim = c(-60,80))
terra::plot(monkey_puzzle, col = 'red', add = TRUE)

#-------------------------------------------------------------------------------

######################################
# 2.1 Range size and range centre ####
######################################

# Range area of alpine shrew in square meters:
# old: 
# area(shrew)

terra::expanse(shrew) # m²

#-------------------------------------------------------------------------------

# Range area of alpine shrew in square kilometers:
# old:
# area(shrew)/1000000

terra::expanse(shrew)/1000000 # km²

#-------------------------------------------------------------------------------

# Range area of monkey puzzle in square meters:
# area(monkey_puzzle)

terra::expanse(monkey_puzzle)

#-------------------------------------------------------------------------------

library(rgeos) # depends on sp and will be retired by the end of 2023

# range centroid:
# old
# gCentroid(shrew)

terra::centroids(shrew)

#-------------------------------------------------------------------------------

# Map the species range and add the centroid to the map
map('world', xlim = c(5,30), ylim = c(40,55))
plot(shrew, col = 'red', add = TRUE)
plot(terra::centroids(shrew), col = 'blue', add = TRUE, lwd = 3)

#-------------------------------------------------------------------------------

# Let's crop the range polygons to South America
# old
# monkey_puzzle_SAm <- gIntersection(monkey_puzzle, as(extent(-85, -30, -55, 5),"SpatialPolygons"))
# 
# # Map the range and range centroid
# map('world',xlim = c(-100,-10),ylim = c(-60,15))
# plot(monkey_puzzle_SAm, col='red',add=T)
# plot(gCentroid(monkey_puzzle_SAm), col='blue',add=T,lwd=3)


monkey_puzzle_SAm <- terra::intersect(monkey_puzzle, ext(-85, -30, -55, 5))

map('world',xlim = c(-100,-10),ylim = c(-60,15))
plot(monkey_puzzle_SAm, col = 'red', add = TRUE)
plot(terra::centroids(monkey_puzzle_SAm), col = 'blue', add = T, lwd = 3)

#-------------------------------------------------------------------------------

#################################
# 2.2 Rasterising range maps ####
#################################
# TODO: doesn't work that way with terra

# By default, terra() will create a 1° resolution map in the *WGS 84* coordinate system (lon/lat).
(r_1deg <- terra::rast())

#-------------------------------------------------------------------------------

(shrew_1deg <- terra::rasterize(shrew, r_1deg))

#-------------------------------------------------------------------------------

map('world',xlim = c(5,30), ylim = c(40,55))
plot(shrew, col = 'red', add = TRUE)
plot(shrew_1deg, add = TRUE, alpha = 0.6, legend = FALSE)

#-------------------------------------------------------------------------------

(r_2deg <- terra::rast(res=2))

# Rasterize the eurasian lynx data
lynx_lynx_2deg <- terra::rasterize(lynx_lynx, r_2deg)

# Map the occupied grid cells
map('world')
plot(lynx_lynx_2deg, add = TRUE, legend = FALSE)


#!!!
# NOTE: I think the next code block about the values can be skipped
#!!! 

#-------------------------------------------------------------------------------

###########################################
# 2.2.2 Rasterising range maps with letsR #
###########################################

library(letsR) # requires raster
library(raster)

# raster version for monkey puzzle (lets.presab doesn't work with terra objects)
(monkey_puzzle_raster <- BIEN_ranges_load_species('Araucaria_araucana'))

# The lets.presab() function expects specific column names in the Polygons data frame
colnames(monkey_puzzle_raster@data) <- "binomial"

# We set the resolution to 1 degree (the default) and restrict the spatial extent to South America
r_monkey_puzzle <- lets.presab(monkey_puzzle_raster, resol = 1, xmn = -100, xmx = -10, ymn = -57, ymx = 15)

# Map the range and range centroid
map('world',xlim = c(-100,-10),ylim = c(-60,15))
plot(monkey_puzzle_SAm,col = 'blue', add = TRUE)
plot(r_monkey_puzzle, add = TRUE, alpha = 0.6, legend = FALSE)

#-------------------------------------------------------------------------------

####################################################################
# 2.2.3 Bulk-rasterising multiple species range maps with letsR ####
####################################################################

# !!!
# NOTE: this will have to be done with raster again
# !!!
# Extract the available Pinus species names
(pinus_names <- BIEN_ranges_genus("Pinus",match_names_only = T)[,1])

#-------------------------------------------------------------------------------

# Download the range maps for all Pinus species
pinus <- BIEN_ranges_load_species(pinus_names)

# Format the column names and rasterise
colnames(pinus@data) <- "binomial"
r_pinus <- lets.presab(pinus, resol = 1, xmn = -170, xmx = -10, ymn = -57, ymx = 60)

# Plot species richness
plot(r_pinus)

#-------------------------------------------------------------------------------

# prepare raster version of the mammals shapefile
mammals_raster <- raster::shapefile('data/MAMMTERR.shp')

# Subset the SpatialPolygonsDataFrame
artibeus_spp <- mammals_raster[grep('Artibeus',mammals_raster@data$BINOMIAL),]

# Rasterize the ranges using the letsR package
r_artibeus_spp <- lets.presab(artibeus_spp, resol = 2, 
                              presence = 1, origin = 1, seasonal = 1)

#-------------------------------------------------------------------------------

# Map the species richness
plot(r_artibeus_spp)

#-------------------------------------------------------------------------------

# Map single species - here, just the first two
par(mfrow=c(1,2))
plot(r_artibeus_spp, name = "Artibeus amplus")
plot(r_artibeus_spp, name = "Artibeus anderseni")

#-------------------------------------------------------------------------------

# Look at structure of the object and at the presence-absence matrix
str(r_artibeus_spp, 1)

#-------------------------------------------------------------------------------

head(r_artibeus_spp$Presence_and_Absence_Matrix)